src/pane.coffee

```diff
@@ -603,55 +605,71 @@ class Pane
   # * `force` (optional) {Boolean} Destroy the item without prompting to save
   #    it, even if the item's `isPermanentDockItem` method returns true.
   #
-  # Returns a {Boolean} indicating whether or not the item was destroyed.
+  # Returns a {Promise} that resolves with a {Boolean} indicating whether or not
+  # the item was destroyed.
   destroyItem: (item, force) ->
     index = @items.indexOf(item)
     if index isnt -1
       return false if not force and @getContainer()?.getLocation() isnt 'center' and item.isPermanentDockItem?()
       @emitter.emit 'will-destroy-item', {item, index}
       @container?.willDestroyPaneItem({item, index, pane: this})
-      if force or @promptToSaveItem(item)
+      if force or not item?.shouldPromptToSave?()
         @removeItem(item, false)
         item.destroy?()
-        true
       else
-        false
+        @promptToSaveItem(item).then (result) =>
+          if result
+            @removeItem(item, false)
+            item.destroy?()
+          result
 
   # Public: Destroy all items.
   destroyItems: ->
-    @destroyItem(item) for item in @getItems()
-    return
+    Promise.all(
+      @getItems().map(@destroyItem.bind(this))
```

Comment:
@maxbrunsfeld 
Noticed while I read code, using `bind` here is bad.
Since `Array.prototype.map` pass item index as 2nd arg, which goes to `@destroyItems` which treat 2nd arg as `force` switch. Which is not intentional I believe.